src/test/java/com/yelp/nrtsearch/server/grpc/PrimaryRestartTests.java

```diff
@@ -110,7 +120,377 @@ public void testPrimaryRestartReplication() throws IOException {
+    primaryServer.addSimpleDocs("test_index", 10);
+    primaryServer.refresh("test_index");
+    primaryServer.addSimpleDocs("test_index", 11);
+    primaryServer.refresh("test_index");
+
+    replicaServer.waitForReplication("test_index");
+    long currentSearcherVersion1 =
+        replicaServer
+            .getClient()
+            .getBlockingStub()
+            .search(
+                SearchRequest.newBuilder()
+                    .setIndexName("test_index")
+                    .setQuery(Query.newBuilder().build())
+                    .build())
+            .getSearchState()
+            .getSearcherVersion();
+    IndexSearcher currentSearcher1 =
+        replicaServer
+            .getGlobalState()
+            .getIndex("test_index")
+            .getShard(0)
+            .slm
+            .acquire(currentSearcherVersion1);
+    List<LeafReaderContext> currentLeaves1 = currentSearcher1.getIndexReader().leaves();
+    replicaServer.getGlobalState().getIndex("test_index").getShard(0).slm.release(currentSearcher1);
+
+    primaryServer.verifySimpleDocIds("test_index", 1, 2, 3, 9, 10, 11);
+    replicaServer.verifySimpleDocIds("test_index", 1, 2, 3, 9, 10, 11);
+
+    verifySimpleDocIdsForVersion(
+        replicaServer, previousSearcherVersion1, "test_index", 1, 2, 3, 4, 5);
+    verifySimpleDocIdsForVersion(
+        replicaServer, previousSearcherVersion2, "test_index", 1, 2, 3, 4, 5, 6, 7, 8);
+
+    primaryServer.addSimpleDocs("test_index", 12);
+    primaryServer.refresh("test_index");
+
+    replicaServer.waitForReplication("test_index");
+    long currentSearcherVersion2 =
+        replicaServer
+            .getClient()
+            .getBlockingStub()
+            .search(
+                SearchRequest.newBuilder()
+                    .setIndexName("test_index")
+                    .setQuery(Query.newBuilder().build())
+                    .build())
+            .getSearchState()
+            .getSearcherVersion();
+    IndexSearcher currentSearcher2 =
+        replicaServer
+            .getGlobalState()
+            .getIndex("test_index")
+            .getShard(0)
+            .slm
+            .acquire(currentSearcherVersion2);
+    List<LeafReaderContext> currentLeaves2 = currentSearcher2.getIndexReader().leaves();
+    replicaServer.getGlobalState().getIndex("test_index").getShard(0).slm.release(currentSearcher2);
+
+    primaryServer.verifySimpleDocIds("test_index", 1, 2, 3, 9, 10, 11, 12);
+    replicaServer.verifySimpleDocIds("test_index", 1, 2, 3, 9, 10, 11, 12);
+
+    verifySimpleDocIdsForVersion(
+        replicaServer, previousSearcherVersion1, "test_index", 1, 2, 3, 4, 5);
+    verifySimpleDocIdsForVersion(
+        replicaServer, previousSearcherVersion2, "test_index", 1, 2, 3, 4, 5, 6, 7, 8);
+
+    verifySegmentsClosed(previousLeaves1);
+    verifySegmentsClosed(previousLeaves2);
+    verifySegmentsClosed(currentLeaves1);
+    verifySegmentsClosed(currentLeaves2);
```

Comment:
When first reading this test without reading the `verifySegmentsClosed` method it appears as if we are checking that all segments in these leaves are closed. Maybe add a comment before this block saying that we're checking that the segments are still open.